chardev/char-pty.c

```diff
@@ -49,64 +48,35 @@ DECLARE_INSTANCE_CHECKER(PtyChardev, PTY_CHARDEV,
 
 static void pty_chr_state(Chardev *chr, int connected);
 
-static void pty_chr_timer_cancel(PtyChardev *s)
-{
-    if (s->timer_src) {
-        g_source_destroy(s->timer_src);
-        g_source_unref(s->timer_src);
-        s->timer_src = NULL;
-    }
-}
-
-static gboolean pty_chr_timer(gpointer opaque)
-{
-    struct Chardev *chr = CHARDEV(opaque);
-    PtyChardev *s = PTY_CHARDEV(opaque);
-
-    pty_chr_timer_cancel(s);
-    if (!s->connected) {
-        /* Next poll ... */
-        qemu_chr_be_update_read_handlers(chr, chr->gcontext);
-    }
-    return FALSE;
-}
-
-static void pty_chr_rearm_timer(Chardev *chr, int ms)
-{
-    PtyChardev *s = PTY_CHARDEV(chr);
-    char *name;
-
-    pty_chr_timer_cancel(s);
-    name = g_strdup_printf("pty-timer-%s", chr->label);
-    s->timer_src = qemu_chr_timeout_add_ms(chr, ms, pty_chr_timer, chr);
-    g_source_set_name(s->timer_src, name);
-    g_free(name);
-}
-
-static void pty_chr_update_read_handler(Chardev *chr)
+static void char_pty_poll_connection_status(Chardev *chr)
 {
     PtyChardev *s = PTY_CHARDEV(chr);
     GPollFD pfd;
     int rc;
     QIOChannelFile *fioc = QIO_CHANNEL_FILE(s->ioc);
 
     pfd.fd = fioc->fd;
-    pfd.events = G_IO_OUT;
+    /* If a PTY is connected, the HUP flag is cleared so
+     * we can poll the connection status this way. */
+    pfd.events = G_IO_HUP;
     pfd.revents = 0;
     rc = RETRY_ON_EINTR(g_poll(&pfd, 1, 0));
     assert(rc >= 0);
 
-    if (pfd.revents & G_IO_HUP) {
-        pty_chr_state(chr, 0);
-    } else {
-        pty_chr_state(chr, 1);
-    }
+    pty_chr_state(chr, !(pfd.revents & G_IO_HUP));
+}
+
+static void pty_chr_update_read_handler(Chardev *chr)
+{
+    char_pty_poll_connection_status(chr);
 }
```

Comment:
What does this function wrapper do?

Reply:
`pty_chr_update_read_handler` is using by QEMU to update the reading side of the device. I am sure a QEMU export can fill the blanks, but the bottom line is that this function was already there and polled the FD. I just factored out the code to `char_pty_poll_connection_status`.